storage/remote/queue_manager.go

```diff
@@ -1058,21 +1073,36 @@ func sendWriteRequestWithBackoff(ctx context.Context, cfg config.QueueConfig, s
 		}
 
 		// If the error is unrecoverable, we should not retry.
-		if _, ok := err.(RecoverableError); !ok {
+		backoffErr, ok := err.(RecoverableError)
+		if !ok {
 			return err
 		}
 
-		// If we make it this far, we've encountered a recoverable error and will retry.
-		onRetry()
-		level.Warn(l).Log("msg", "Failed to send batch, retrying", "err", err)
-
-		time.Sleep(time.Duration(backoff))
-		backoff = backoff * 2
-
-		if backoff > cfg.MaxBackoff {
-			backoff = cfg.MaxBackoff
+		sleepDuration = backoff
+		if !backoffErr.retryAfter.isDateTime {
+			if backoffErr.retryAfter.duration != noRetryAfter {
+				sleepDuration = backoffErr.retryAfter.duration
+			}
+			sleep(backoffErr)
+		} else {
+			// date-time is used for Retry-After. Hence, we need to wait here until we cross the provided
+			// retry-after date-time.
+			waitUntil := backoffErr.retryAfter.after
+			if waitUntil.After(time.Now()) {
+				checkCurrentTimeInterval := time.NewTimer(time.Minute)
+				for {
+					currentTime := <-checkCurrentTimeInterval.C
+					if currentTime.After(waitUntil) {
+						// We have crossed the wait duration provided by retry-after. Lets attempt again.
+						break
+					}
+				}
+				onRetry()
+			} else {
+				level.Info(l).Log("msg", "retry-after cannot be in past, retrying the default backoff way")
```

Comment:
I think this could be a debug log. I would expect it to happen sometimes and it is a reasonable behavior. You could also get rid of it by always sleeping the larger amount of `backoff` or `retryAfter`.